flashlight/fl/tensor/backend/jit/eval/Evaluator.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/eval/Evaluator.h"
+
+#include "flashlight/fl/tensor/backend/jit/JitTensorBase.h"
+#include "flashlight/fl/tensor/backend/jit/ir/ValueNode.h"
+
+namespace fl {
+
+namespace {
+
+void fillNodeRefCountInTree(
+    Node* root,
+    std::unordered_map<Node*, unsigned>& nodeToRefCount) {
+  if (nodeToRefCount.find(root) == nodeToRefCount.end()) {
+    nodeToRefCount.emplace(root, root->getRefCount());
+    for (const auto& input : root->inputs()) {
+      fillNodeRefCountInTree(input, nodeToRefCount);
+    }
+  }
+}
+std::unordered_map<Node*, unsigned> getNodeToRefCountInTree(Node* root) {
+  std::unordered_map<Node*, unsigned> nodeToRefCount;
+  fillNodeRefCountInTree(root, nodeToRefCount);
+  return nodeToRefCount;
+}
+
+} // namespace
+
+Evaluator::Evaluator(TensorBackend& backend) : backend_(backend) {}
+
+void Evaluator::evalBinaryNode(BinaryNode& node) {
+  evalNode(node.lhs());
+  evalNode(node.rhs());
+  const auto& lhs = node.lhs()->getResult().value();
+  const auto& rhs = node.rhs()->getResult().value();
+  node.setResult(evalBinaryOp(node.op(), lhs, rhs));
+}
+
+void Evaluator::evalCustomNode(CustomNode& node) {
+  std::vector<const Tensor*> inputTensors;
+  for (auto& inputNode : node.inputs()) {
+    evalNode(inputNode);
+    inputTensors.push_back(&inputNode->getResult().value());
+  }
+  node.setResult(node.evalFunc()(inputTensors));
+}
+
+void Evaluator::evalScalarNode(ScalarNode& node) {
+  node.setResult(evalScalar(node));
+}
+
+Tensor
+Evaluator::evalBinaryOp(BinaryOp op, const Tensor& lhs, const Tensor& rhs) {
+  switch (op) {
+    case BinaryOp::Add:
+      return backend_.add(lhs, rhs);
+    case BinaryOp::Sub:
+      return backend_.sub(lhs, rhs);
+    case BinaryOp::Mul:
+      return backend_.mul(lhs, rhs);
+    case BinaryOp::Div:
+      return backend_.div(lhs, rhs);
+  }
+  throw std::runtime_error(
+      "[Evaluator::evalBinaryOp] Unknown binary operation type");
+}
+
+Tensor Evaluator::evalScalar(ScalarNode& node) {
+  const Shape& shape = node.shape();
+  const auto dtype = node.dataType();
+  switch (dtype) {
+    case dtype::f16:
+    case dtype::f32:
+    case dtype::f64:
+      return backend_.full(shape, node.scalar<double>(), dtype);
+    case dtype::b8:
+    case dtype::s16:
+    case dtype::s32:
+    case dtype::s64:
+    case dtype::u8:
+    case dtype::u16:
+    case dtype::u32:
+      return backend_.full(shape, node.scalar<long long>(), dtype);
```

Comment:
Some of these widen pretty massively -- is that just to simplify the implementation?

Reply:
> just to simplify the implementation?

Yes. I could use a variant that holds each possible type, but since `long long`, `unsigned long long` and `double` could capture all types w/o loss of precision, I went with these 3.

Reply:
Sure -- I guess this only matters for the memory footprint of `Node` and not an actual tensor...